File: src/main/java/dev/galiev/sds/SpawnDimensionSetter.java
package dev.galiev.sds;

import com.mojang.logging.LogUtils;
import dev.galiev.sds.config.ConfigManager;
import dev.galiev.sds.helper.WorldHelper;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.event.server.ServerStartingEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import org.slf4j.Logger;

@Mod(SpawnDimensionSetter.MOD_ID)
public class SpawnDimensionSetter {
    public static final String MOD_ID = "sds";
    public static final Logger LOGGER = LogUtils.getLogger();

    public SpawnDimensionSetter() {
        var modEventBus = FMLJavaModLoadingContext.get().getModEventBus();
        MinecraftForge.EVENT_BUS.register(this);
        ConfigManager.init();

/*        ServerLifecycleEvents.SERVER_STARTED.register((MinecraftServer server) -> {
            server.getWorlds().forEach(world -> putWorld(world.getRegistryKey().getValue()));
        });
        CommandRegistrationCallback.EVENT.register((dispatcher, dedicated) -> {
            WorldsCommands.register(dispatcher);
        });*/
    }

    @SubscribeEvent
    public void onServerStarting(ServerStartingEvent event) {
        event.getServer().getAllLevels().forEach(WorldHelper::putWorld);
    }
}

File: src/main/kotlin/dev/galiev/sds/SpawnDimensionSetter.kt
package dev.galiev.sds

import dev.galiev.config.ConfigManager
import dev.galiev.helper.WorldHelper
import net.minecraft.core.BlockPos
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerLevel
import net.minecraft.stats.Stats
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.event.entity.EntityJoinLevelEvent
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.common.Mod
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

@Mod(SpawnDimensionSetter.ID)
object SpawnDimensionSetter {
    const val ID = "sds"

    val LOGGER: Logger = LogManager.getLogger(ID)

    init {
        MinecraftForge.EVENT_BUS.register(this)
        ConfigManager
    }

    @Mod.EventBusSubscriber(modid = ID)
    class ForgeEvents {
        companion object {
            @SubscribeEvent
            fun onPlayerJoin(event: EntityJoinLevelEvent) {
                val server = event.level.server
                LOGGER.info("Server started")
                server?.playerList?.players?.forEach { serverPlayer ->
                    if (serverPlayer.stats.getValue(Stats.CUSTOM.get(Stats.LEAVE_GAME)) < 1) {
                        val level: ServerLevel = ConfigManager.read().dimension.split(":").let { value ->
                            server.allLevels.find { it.dimension().location() == ResourceLocation(value[0], value[1]) }
                        } ?: return server.close()
                        val x = WorldHelper.getRandInt(ConfigManager.read().rangeX)
                        val y = 70
                        val z = WorldHelper.getRandInt(ConfigManager.read().rangeZ)
                        val pos = BlockPos.MutableBlockPos(x, y, z)
                        if (WorldHelper.safeCheck(level, pos)) {
                            WorldHelper.tpSafeZone(serverPlayer, level, pos)
                        } else {
                            LOGGER.warn("Safe position doesn't exists")
                        }
                    } else {
                        LOGGER.warn("Player $serverPlayer already sign up")
                    }
                }
            }
        }
    }
}